File: leetcode/904 Fruit Into Baskets/solution.go
func totalFruit(fruits []int) int { // 滑动窗口虫取法！！！
    hashRecord := make(map[int]int)
    k, left, right := 2, 0, 0
    for right < len(fruits) {
        _, ok := hashRecord[fruits[right]]
        if ok {
            hashRecord[fruits[right]]++
        } else {
            hashRecord[fruits[right]] = 1
        }
        if len(hashRecord) > k {
            hashRecord[fruits[left]]--
            if hashRecord[fruits[left]] == 0 {
                delete(hashRecord, fruits[left])
            }
            left++
        }
        right++
    }
    return right-left
}

File: leetcode/904 Fruit Into Baskets/solution.java
class Solution {
    public int totalFruit(int[] fruits) { // 滑动窗口虫取法！！！
        Map<Integer, Integer> hashRecord = new HashMap<>();
        int k = 2;
        int left = 0, right = 0;
        while (right < fruits.length) {
            hashRecord.put(fruits[right], hashRecord.getOrDefault(fruits[right], 0)+1);
            if (hashRecord.size() > k) {
                hashRecord.put(fruits[left], hashRecord.get(fruits[left])-1);
                if (hashRecord.get(fruits[left]) == 0) hashRecord.remove(fruits[left]);
                left++;
            }
            right++;
        }
        return right-left;
    }
}